src/patchkernel/patch_kernel.tpp

```diff
@@ -45,6 +45,150 @@ std::unique_ptr<patch_t> PatchKernel::clone(const patch_t *original)
+	// ids of cells to be deleted contains the last internal cell, we delete
+	// that cell ater deleting all other cells. In this way we make sure to
+	// deleting the last internal cells just once (after deleting the last
+	// internal, another cells becomes the last one and  that cells may be
+	// on the deletion list, and on and so forth). The same applies for the
+	// first ghost.
+	bool deleteLastInternalCell = false;
+#if BITPIT_ENABLE_MPI==1
+	bool deleteFirstGhostCell = false;
+#endif
+	for (long id : ids) {
+		if (id == m_lastInternalCellId) {
+			deleteLastInternalCell = true;
+			continue;
+		}
+#if BITPIT_ENABLE_MPI==1
+		else if (id == m_firstGhostCellId) {
+			deleteFirstGhostCell = true;
+			continue;
+		}
+#endif
+
+		deleteCell(id);
+	}
+
+	if (deleteLastInternalCell) {
+		deleteCell(m_lastInternalCellId);
+	}
+
+#if BITPIT_ENABLE_MPI==1
+	if (deleteFirstGhostCell) {
+		deleteCell(m_firstGhostCellId);
+	}
+#endif
+
+	return true;
+}
+
+/*!
+	Deletes a list of vertices.
+
+	\param ids are the ids of the vertices to be deleted
+*/
+template<typename IdStorage>
+bool PatchKernel::deleteVertices(const IdStorage &ids)
+{
+	if (!isExpert()) {
+		return false;
+	}
+
+	// Deleting the last vertex requires some additional work. If the ids
+	// of vertices to be deleted contain the last vertex, that vertex is
+	// deleted after deleting all other vertices. In this way we make sure
+	// to delete the last vertex just once (after deleting the last vertex,
+	// another vertex becomes the last one and that vertex may be on the
+	// deletion list as well, and on and so forth). The same applies for
+	// the first ghost.
+	bool deleteLastInternalVertex = false;
+#if BITPIT_ENABLE_MPI==1
+	bool deleteFirstGhostVertex = false;
+#endif
+	for (long id : ids) {
+		if (id == m_lastInternalVertexId) {
+			deleteLastInternalVertex = true;
+			continue;
+		}
+#if BITPIT_ENABLE_MPI==1
+		else if (id == m_firstGhostVertexId) {
+			deleteFirstGhostVertex = true;
+			continue;
+		}
+#endif
+
+		deleteVertex(id);
+	}
+
+	if (deleteLastInternalVertex) {
+		deleteVertex(m_lastInternalVertexId);
+	}
+
+#if BITPIT_ENABLE_MPI==1
+	if (deleteFirstGhostVertex) {
+		deleteVertex(m_firstGhostVertexId);
+	}
+#endif
+
+	return true;
+}
+
+/*!
+	Deletes a list of interfaces.
+
+	\param ids are the ids of the interfaces to be deleted
+*/
+template<typename IdStorage>
+bool PatchKernel::deleteInterfaces(const IdStorage &ids)
+{
+	if (!isExpert()) {
+		return false;
+	}
+
+	// Deleting the last interface requires some additional work. If the ids
+	// of interfaces to be deleted contain the last interface, that interface
+	// is deleted after deleting all other interfaces. In this way we make
+	// sure to delete the last interface just once (after deleting the last
+	// interface, another interface becomes the last one and that interface
+	// may be on the deletion list as well, and on and so forth).
+	std::size_t lastId;
```

Comment:
think this lastId should be of type `long` / `id_t`

Reply:
Will fix it, thanks.

Reply:
Fixed.